20240129-satellite-managed-encryption.md

```diff
@@ -0,0 +1,164 @@
+
+Additionally, by default, object paths are encrypted. This has privacy benefits, but prevents certain features that some users are interested in, like lexicographic ordering of object keys. Like with user-managed encryption passphrases, we want to continue supporting path encryption for users who are interested, while allowing other users who want lexicographic ordering features to create a project which has path encryption turned off.
+
+### Goals
+
+Users are provided the option to create a project that uses "satellite-managed encryption". One of these projects:
+* uses a securely-stored, random passphrase stored on the satellite as the project's encryption passphrase
+* restricts all files to be uploaded without path encryption
+
+The main goals are to:
+* provide users with an experience that does not require managing and entering an encryption passphrase in the Satellite UI
+* provide users with an easy way to view and list files in lexographical order
+
+### Approach / Design
+
+#### Key Management Store
+
+We need a secure way to store and look up project encryption passphrases on the Satellite. Someone with normal Satellite DB access should not be able to view this data, and ideally it is encrypted. At the same time, it must be very reliable, as losing data about project encryption passphrases means that data may have been lost. The primary features and concerns we should keep in mind:
+* as simple as possible from an implementation perspective, e.g. key-value lookup for project ID -> encryption passphrase
+* secure - should be encrypted, and highly restricted who/what has access to it
+* reliable - should have the same standards for uptime/data loss as metainfo DB
+
+The KMS will be made up of 3 components:
+* The master key and store - The master key will be used to encrypt and decrypt passphrases. The store will be used to securely store the master key. The store will be [Google's Secret Manager](https://cloud.google.com/security/products/secret-manager), which is a secure store for secrets such as API keys and other sensitive data.
+This secrets manager was chosen because it is just as secure as the alternatives, if not more and yet cheaper and easier to manage. It also is more convenient to use as we already use GCP.
+* The KMS service - This is part of the Satellite that will be responsible for fetching the master key at startup, encrypting and decrypting passphrases, and storing and retrieving same from the passphrase store. The master key will be kept in memory throughout the lifetime of the Satellite.
+The satellite will be configured with credentials needed to access the Secret Manager (the master key store).
+* The Passphrase store - This where the encrypted passphrases will be stored. It will be a new column, `passphrase_enc` on the `projects` table. As added benefits to storing this on `projects`, we will be able to back up encrypted passphrases and also rotate master keys.
```

Comment:
nit - I don't think we need to think of "the passphrase store" as its own thing. Above, we have "The KMS service ... will be responsible for ... encrypting and decryptiong passphrases, and storing and retrieving from the passphrase store"

I think the KMS maybe doesn't even need to be thought of as a "store", but rather just a simple "encryption/decryption service", whose primary responsibilities are:
* keep the master key in memory
* expose, e.g. `managedPassphraseService.GenerateNewPassphrase()` and `managedPassphraseService.Decrypt(encryptedPassphrase)` functionality

In other words, the "managed passphrase service" doesn't even need to have DB access - we are storing `passphrase_enc` in the DB, and this encrypted passphrase will already be accessible from the satellite/console package - all we need to be able to do is decrypt it, which is a very simple operation.

One reason I think this slight change would help:
Imagine we are creating a new project - we call db.CreateProject and pass in all the initial field values. This is happening from the Console service. That means that ideally, the Console service needs to have the encrypted project passphrase so that it can insert `passphrase_enc` at the same time the project is created. Otherwise, project creation and setting the satellite managed encryption passphrase need to happen as two separate DB queries.